README.md
### 前言

这是一个 17 年底, 因为无聊做的一个 自己用的博客, 想通过一条龙的方式,学习前端,后端,和部署

最初部署在 阿里云上的,一年后到期了,因为穷,我就改用 `hexo`, 想看看我的新博客吗 [点这里](https://www.lijinke.cn)

在这里分享给大家, 当然,代码写的很糟糕,

如果你想知道他是怎么搭建的 请看 [李金珂的小屋诞生记](https://www.lijinke.cn/2017/05/22/%E6%9D%8E%E9%87%91%E7%8F%82%E7%9A%84%E5%B0%8F%E5%B1%8B%E8%AF%9E%E7%94%9F%E8%AE%B0/)

如果你喜欢的话, 就随意折腾吧. 只需保留下图的 注释即可 : )

![https://cdn.lijinke.cn/WX20190522-150913@2x.png](https://cdn.lijinke.cn/WX20190522-150913@2x.png)

[GITHUB](https://github.com/lijinke666/jinke-blog)

### 预览

> 炫酷的动画 (当时自己想的创意)

![https://cdn.lijinke.cn/ezgif.com-resize.gif](https://cdn.lijinke.cn/ezgif.com-resize.gif)

> 文章页

![https://cdn.lijinke.cn/20180518151903.png](https://cdn.lijinke.cn/20180518151903.png)

> 可以直接上传文章

![https://cdn.lijinke.cn/20180518151903.png](https://cdn.lijinke.cn/WX20190522-153529@2x.png)

>关于页

![https://cdn.lijinke.cn/WX20190522-150749@2x.png](https://cdn.lijinke.cn/WX20190522-150749@2x.png)

> 性感的播放器

![https://cdn.lijinke.cn/WX20190522-153828@2x.png](https://cdn.lijinke.cn/WX20190522-153828@2x.png)

![https://cdn.lijinke.cn/WX20190522-153844@2x.png](https://cdn.lijinke.cn/WX20190522-153844@2x.png)

### 大致的功能

就像上图所示主要就 5 个功能

- 在线聊天 (socket.io)
- 相册,放一些你喜欢的图片
- 文章 (markdown 文章, 和评论)
- 关于,写一些你的介绍
- 音乐播放器

### 使用技术栈

> 没错 年代后久远, 技术栈的版本比较旧
- `React v16.X`
- `ES6,ES7`
- `webpack v3.0.0`
- `Redux`
- `React-Redux`
- `React-Router v3.x`
- `node.js`
- `yarn`
- `mongodb`
- `mongoose`
- `mocha`
- `chai`
- `less`
- `socket.io`
- `pwa`
- `web push`

### node环境
- `v8.9+`

- 安装依赖

```
yarn
```
- 跑起来
```
yarn start
```
- 打包
```
yarn run build
```
- 开发server
```
yarn run dev-server
```
- 生产server
```
yarn run prod-server
```
- 连接数据库
```
yarn run connect-db
```
- 备份数据库
```
yarn run dump
```
- 写入数据库
```
yarn run restore
```

### 本地运行步骤

1. 克隆项目 `git clone https://github.com/lijinke666/jinke-blog.git`
2. 请确保你本地已经安装 `mongodb`, `node.js`, `nodemon` 和 `pm2`
3. 更改 `config/index.js` 的 `你自己的相关配置信息`
3. 切换到目录 `cd jinke-blog` 连接数据库 `yarn run connect-db`
4. 运行前端 `yarn start` 等待打包 自动打开浏览器 `localhost:6688`
5. 运行后端 `yarn run dev-server`


### 数据库

> 表格当时瞎设计的 就三张表格

```js
const mongoose = require('mongoose')
const debug = require('debug')('music-schema')
const Schema = mongoose.Schema

const musicSchema = new Schema({
    name: String,           //音乐名
    src: String,            //路径
    cover: String,          //图片路径
    desc:String ,            //描述
    isShow: {                // 发布文章后 默认需要审核 你可以将它改为 false 直接查看
      type:Boolean,
      default:true
    }
}, {
        collection: "music"
    })

const articleSchema = new Schema({
    title: String,           //文章标题
    content: String,          //文章内容   markdown 格式
    previewContent:String,     //预览的文章内容
    author: String,           //作者
    publishDate: {            //发表日期
        type:Date,
        default:Date.now
    },
    pageView: Number,         //点击量
    like: Number,             //喜欢数量
    approve:Boolean,           //是否审核通过
    email:String,              //作者邮箱  用来通知作者文章是否通过
    category:Array,
}, {
        collection: "article"
    })

const commentSchema = new Schema({
        articleId:String,           //文章id
        commentName:String,         //姓名
        commentEmail:String,        //邮箱
        commentContent:String,      //内容
        like:String,                //点赞量
        device:String,               //设备
        publishDate:{               //发布日期
            type:Date,
            default:Date.now
        }
},{
    collection:"comment"
})
```


CHANGELOG.md
# V2.2.0 
> 2017/12/6 

---
## Features
- 关于我们增加分享功能
- 关于我们增加有意思的语音
- `Button` 组件 字体大小 调整为 `14` => `16`
- `Modal` 组件 优化动画 淡入淡出, 调整间距,使其更扁平化
- `MusicPlayer`组件 增加几首新歌曲
- `Aritcle` 页面 增加 文章搜索功能
- `ArticleDetail` 文章详情页 优化动画,更平滑
- `Message` 组件 优化动画
- 服务端 api 请求超时 设为 10s

## Bug Fixs
- `fix` 文章页码 切换时 无动画 bug
- `fix` 排行榜 切换时 无动画 bug
- `fix` 文章详情页 评论数 bug
- `fix` reducer 默认值 的 bug



# V2.2.1
> 2018/3/21
## Features
- 增加 返回顶部 的功能
- 更换关于我们图片

## Bug Fixs
- 修复文章页刷新404 bug


# V2.3.0
> 2018/3/27
## Features
- 增加 service worker 缓存
- 增加 PWA 体验
- 增加 文章订阅消息推送
- 控制台新增打印 版本号

# V2.4.0
> 2018/4/27
- 修复网站刷新bug,重定向到index.html
- 文章增加分享
- 更新 `react-jinke-music-player` 播放器版本
- 新用户显示 欢迎页 老用户去掉
- 增加404 页面

# V2.5.0
> 2018/5/16
- 升级 `babel7`
- webpack 版本 使用 2.7.0
- 播放器版本 更新 3.4.2
- css 打包成多个css 
- 升级 `uglifyjs-webpack-plugin` 版本 修复 压缩js 失败的问题

# V2.5.1
> 2018/5/18
- 去掉 html5 操蛋 的 `manifest`
- server 设置 `Cache-Controller` 缓存
- 调整播放器初始位置

# V2.6.0
> 2018/5/24
- 升级 `HTTPS`
- 支持 `PWA`
- 网站初始化增加 `Loading`

# V2.7.0
> 2018/7/25
- 优化 `PWA`
- 增加 1周年庆活动
- 优化 `Modal` `Button` 组件 调整尺寸
- 增加 1周年庆活动相应接口
- 评论文章 增加通知


# V2.8.0
> 2018/8/7
- 优化 `PWA`
- 增加 `services worker`
- express 服务器相关 改为 `nginx` 配置
- 迁移服务器
- public/images 是一周年存放图片的地址
- public/logos 是 pwa 所需 icon的地址
- public/music 是音乐播放文件


webpack.config.js
/*
 * @Author: jinke.li 
 * @Date: 2017-05-03 16:32:21 
 * @Last Modified by: jinke.li
 * @Last Modified time: 2018-08-22 21:29:59
 */
const path = require("path");
const webpack = require("webpack");
const HtmlWebpackPlugin = require("html-webpack-plugin"); //自动生成一个html 引入打包之后的js
const ExtractTextPlugin = require("extract-text-webpack-plugin"); //默认打包css 这些全部在js 里面  用这个可以分离出来 单独生成css文件  //生产环节会用到
const OpenBrowserPlugin = require("open-browser-webpack-plugin"); //打包完成自动打开浏览器
const CopyWebpackPlugin = require("copy-webpack-plugin"); //拷贝文件  当有第三方依赖可以copy到打包文件夹中
const CptimizeCssAssetsPlugin = require("optimize-css-assets-webpack-plugin"); //压缩css
const ImageminPlugin = require("imagemin-webpack-plugin").default; //压缩图片
const UglifyJSPlugin = require("uglifyjs-webpack-plugin"); //webpack3 单独分离出来了这个压缩的
const SWPrecacheWebpackPlugin = require("sw-precache-webpack-plugin"); //生成 service-worker 配置文件  妈的以前自己写...

const { host, dev_port } = require("./config");

module.exports = env => {
  //env 是npm script 运行webpack时传进来的  判断是否是开发环境
  const mode = (env && env.mode) || "development";

  const options = {
    //开发工具
    devtool: mode === "development" ? "source-map" : false,

    //开发服务器
    devServer: {
      contentBase: [
        path.resolve(__dirname, "dist"),
        path.resolve(__dirname, "public")
      ], //静态资源根目录
      compress: true, //压缩
      port: dev_port, //端口
      hot: true, //热更新
      inline: true, //iframe 模式
      historyApiFallback: true, //浏览器 history
      stats: {
        //统计
        color: true, //输出有颜色的信息
        errors: true, //显示错误信息
        version: true, //显示版本号
        warnings: true, //显示警告
        progress: true, //显示进度,
        timings: true //显示时间
      }
    },

    //入口
    entry:
      mode === "development"
        ? [
            "react-hot-loader/patch", //热更新
            `webpack-dev-server/client?${host}:${dev_port}`,
            "webpack/hot/only-dev-server",
            path.resolve(__dirname, "src/index.js")
          ]
        : {
            app: path.resolve(__dirname, "src/index.js")
            // vendor:['react']
          },

    //打包输出
    output: {
      path: path.resolve(__dirname, "dist"),
      filename:
        mode === "development" ? "js/[name].js" : "js/[name].[chunkhash:8].js",
      chunkFilename:
        mode === "development" ? "js/[name].js" : "js/[name].[chunkhash:8].js",
      publicPath: mode === "development" ? `${host}:${dev_port}/` : "/"
    },

    //模块加载器
    module: {
      rules: [
        {
          test: /\.js[x]?$/,
          use: [
            {
              loader: "babel-loader"
            }
          ],
          exclude: "/node_modules/",
          include: [path.resolve("src")] //只遍历src目录下的
        },
        {
          test: /\.less$/,
          use:
            mode === "development" //开发环境 css打包到js中
              ? [
                  { loader: "style-loader" }, //loader 倒序执行  先执行 less-laoder
                  {
                    loader: "css-loader",
                    options: { minimize: false, sourceMap: true }
                  },
                  { loader: "postcss-loader" }, //自动加前缀
                  { loader: "less-loader", options: { sourceMap: true } }
                ]
              : ExtractTextPlugin.extract({
                  //生产环境 把css单独分离出来
                  fallback: "style-loader",
                  use: [
                    "css-loader",
                    "postcss-loader",
                    {
                      loader: "less-loader",
                      options: {
                        sourceMap: false
                      }
                    }
                  ]
                })
        },
        {
          test: /\.css$/,
          use:
            mode === "development"
              ? [
                  { loader: "style-loader" }, //loader 倒序执行  先执行 less-laoder
                  {
                    loader: "css-loader",
                    options: { minimize: false, sourceMap: true }
                  },
                  { loader: "postcss-loader" }
                ]
              : ExtractTextPlugin.extract({
                  fallback: "style-loader",
                  use: [
                    "css-loader",
                    "postcss-loader",
                    {
                      loader: "less-loader",
                      options: {
                        sourceMap: false
                      }
                    }
                  ]
                })
        },
        {
          test: /\.(jpg|jpeg|png|gif|cur|ico)$/,
          use: [
            {
              loader: "file-loader",
              options: {
                name: "images/[name][hash:8].[ext]" //遇到图片  生成一个images文件夹  名字.后缀的图片
              }
            }
          ]
        },
        {
          test: /\.(eot|ttf|svg|woff|woff2)$/,
          use: [
            {
              loader: "file-loader",
              options: {
                name: "fonts/[name][hash:8].[ext]"
              }
            }
          ]
        }
      ]
    },

    //自动补全后缀
    resolve: {
      enforceExtension: false, //2.0 后 不能写 extensions :[""]
      extensions: [".js", ".jsx", ".json"], //比如 test.js   可以写成 require('test')
      modules: [
        path.resolve("src"), //比如 src/app/components/xx  可以写成 app/components/xx
        path.resolve("."),
        path.resolve("src/shared"),
        "node_modules"
      ]
    },

    //插件
    plugins: []
  };
  //根据开发环境不同  concat 不同的插件
  if (mode === "development") {
    options.plugins = options.plugins.concat([
      new webpack.NamedModulesPlugin(), //打印更具可读性模块名称在浏览器控制台
      new webpack.NoEmitOnErrorsPlugin(), //错误不打断
      new webpack.DefinePlugin({
        //调试
        __DEBUG__: true
      }),
      new webpack.HotModuleReplacementPlugin(), //热加载插件
      new OpenBrowserPlugin({
        //编译完成打开浏览器
        url: `${host}:${dev_port}`
      })
    ]);
  } else {
    options.plugins = options.plugins.concat([
      // new BundleAnalyzerPlugin(),     //生成打包图
      // //webpackv3.0新增 作用域提升 默认是闭包式打包 浏览器执行速度变慢
      // //开启这个去掉模块的包裹函数,体积更小
      // new webpack.optimize.ModuleConcatenationPlugin(),
      new webpack.HashedModuleIdsPlugin(), //生成稳定的hashId 没有改变的chunk文件这样hash不会变
      new webpack.DefinePlugin({
        "process.env.NODE_ENV": JSON.stringify("production"),
        __DEBUG__: false
      }),
      new UglifyJSPlugin(),
      new ExtractTextPlugin({
        // 将打包文件中的css分离成一个单独的css文件
        filename: "css/[name].[chunkhash:8].css",
        allChunks: true
      }),
      //[1]
      //找到所有node_modules的依赖包  分离出来
      // /axios/ 没有用到的模块
      new webpack.optimize.CommonsChunkPlugin({
        name: "app",
        async: "common-in-lazy",
        children: true,
        minChunks: ({ resource } = {}) =>
          resource &&
          resource.includes("node_modules") &&
          /axios/.test(resource)
      }),
      // [2]
      //找到模块次数使用两次的  分离出来
      //单独打成used-twice.js 减少包的体积
      /**
       * 升级到 v2.6 貌似async不起作用  article admin detail 都使用了但是moment都打包进了对应的chunk文件
       * 导致文件增大了600kb
       * 经过github上的提问 各路大神的帮助下  解决了上面这个问题 需要设置name!!!!!!!!!!!
       */
      new webpack.optimize.CommonsChunkPlugin({
        name: "app",
        children: true,
        async: "used-twice",
        minChunks: (module, count) => count >= 2
      }),
      //[3]
      //[1][2][3] 是按需加载 大幅减少打包js体积的关键
      //遍历node_modules目录 以.js结尾 一道vender chunk
      //自动化分离第三方依赖
      new webpack.optimize.CommonsChunkPlugin({
        name: "app",
        filename: "js/common.[chunkhash:8].js",
        minChunks: ({ resource }) =>
          resource &&
          resource.indexOf("node_modules") >= 0 &&
          resource.match(/\.js$/)
      }),
      new webpack.optimize.CommonsChunkPlugin({
        name: ["manifast"]
      }),
      new webpack.LoaderOptionsPlugin({
        //laoder最小化
        minimize: true
      }),
      //图片压缩没用。。。什么鬼
      new ImageminPlugin({
        // disable:false,
        test: /\.(jpe?g|png|gif|svg)$/i,
        optipng: {
          optimizationLevel: 7
        }
      }),
      new CptimizeCssAssetsPlugin({
        //压缩css  与 ExtractTextPlugin 配合使用
        cssProcessor: require("cssnano"),
        cssProcessorOptions: { discardComments: { removeAll: true } }, //移除所有注释
        canPrint: true //是否向控制台打印消息
      })
      //这个插件是我自己写的  用来动态生成webpack打包之后 的cache文件
      // new AddStaticCachePlugin({
      //     template:path.resolve(__dirname,'cacheTemp.tpl'),
      //     cacheName:"lijinkeWeb.appcache",            //缓存文件名
      //     comments:"lijinkeWeb appCache by add-static-cache-webpack-plugin",       //注释
      //     publicPath:"/"                 //公共路径
      // })
    ]);
  }
  options.plugins.push(
    new HtmlWebpackPlugin({
      title: "李金珂的小屋",
      filename: "index.html", //自动把打包的js文件引入进去
      template: path.resolve(__dirname, "src/index.html"), //模板文件
      hash: false, //添加hash码
      inject: true //注射所有资源到 body元素的底部     "head" "body" true false  "body" == true
    }),
    new CopyWebpackPlugin([
      {
        from: path.resolve(__dirname, "manifest.json"),
        to: path.resolve(__dirname, "dist/manifest.json")
      }
    ]),
    new SWPrecacheWebpackPlugin({
      cacheId: "lijinkeWeb",
      dontCacheBustUrlsMatching: false,
      filename: "sw.js",
      minify: false,
      navigateFallback: host + "/index.html",
      staticFileGlobsIgnorePatterns: [/\.map$/, /manifest\.json$/],
      importScripts: [
          path.resolve(__dirname,"src/shared/libs/webpush-sw.js")
      ]
    })
  );
  return options;
};
